packages/components/src/components/select/select/select.tsx

```diff
@@ -0,0 +1,391 @@
+            id={buttonId}
+            {...getComboboxAriaAttributes(this.isOpen, this.required, labelId, ariaDescribedBy, dropdownId)}
+            disabled={this.disabled}
+            onClick={this.onComboClick}
+            onKeyDown={this.onComboKeyDown}
+            ref={(el) => (this.combobox = el)}
+          >
+            {getSelectedOptionString(this.selectOptions)}
+          </button>
+          <PrefixedTagNames.pIcon
+            class={{ icon: true, 'icon--rotate': this.isOpen }}
+            name="arrow-head-down"
+            theme={this.theme}
+            color={this.disabled ? 'state-disabled' : 'primary'}
+            aria-hidden="true"
+          />
+          <div
+            id={dropdownId}
+            class="listbox"
+            {...getListAriaAttributes(this.label, this.required, false, this.isOpen)}
+            tabindex="-1"
+            ref={(el) => (this.listElement = el)}
+          >
+            <slot />
+          </div>
+        </div>
+        <StateMessage state={this.state} message={this.message} theme={this.theme} host={this.host} />
+        <span class="sr-only" role="status" aria-live="assertive" aria-relevant="additions text">
+          {this.srHighlightedOptionText}
+        </span>
+        {this.isWithinForm && <slot name={INTERNAL_SELECT_SLOT} />}
+      </div>
+    );
+  }
+
+  private onSlotchange = (): void => {
+    this.updateOptions();
+    updateSelectOptions(this.selectOptions, this.value);
+    if (this.isWithinForm) {
+      updateNativeSelectOption(this.nativeSelect, this.selectOptions);
+    }
+    // Necessary to update selected options in placeholder
+    forceUpdate(this.host);
+  };
+
+  private updateOptions = (): void => {
+    this.selectOptions = Array.from(this.host.children).filter(
+      (el) => el.tagName !== 'SELECT' && el.slot !== 'label' && el.slot !== 'description' && el.slot !== 'message'
+    ) as HTMLPMultiSelectOptionElement[];
+    this.selectOptions.forEach((child) => throwIfElementIsNotOfKind(this.host, child, 'p-select-option'));
+  };
+
+  private updateSelectedOption = (selectedOption: SelectOption) => {
+    // option can be undefined when no option is highlighted and keyboard action calls this
+    if (selectedOption) {
+      this.preventOptionUpdate = true; // Avoid unnecessary updating of options in value watcher
+      setSelectedOption(this.selectOptions, selectedOption);
+      this.value = selectedOption.value;
+      this.emitUpdateEvent();
+      this.updateSrHighlightedOptionText();
+    }
+    this.updateMenuState(false);
+    this.combobox.focus();
+  };
+
+  private onComboClick = (): void => {
+    this.updateMenuState(!this.isOpen);
+  };
+
+  private updateMenuState = (open: boolean): void => {
+    if (this.isOpen === open) {
+      return;
+    }
+    this.isOpen = open;
+  };
+
+  private onComboKeyDown = (event: KeyboardEvent): void => {
+    const { key } = event;
+
+    const action = getActionFromKey(event, this.isOpen);
```

Comment:
Perhaps rename to `getActionFromKeyboardEvent` since this is how you call function based on parameters. Currently it is a bit awkward that `key` is destructured beforehand but then not passed to this function.